File: test/manifest.test.js
'use strict';

const assert = require('node:assert/strict');
const test = require('node:test');
const {
  buildInputSchema,
  manifestMatchesLocation,
  validateManifest,
} = require('../extension/manifest-contract');

const locator = (css = null) => ({
  css,
  role: null,
  name: null,
  placeholder: null,
  text: null,
  hrefContains: null,
});

const adapter = () => ({
  schemaVersion: 'learned-adapter/1',
  usable: true,
  site: {
    origin: 'https://example.com',
    routePatterns: ['/search'],
  },
  tool: {
    name: 'search_products',
    description: 'Search products',
    safety: 'read',
    parameters: [{
      name: 'query',
      description: 'Search query',
      type: 'string',
      required: true,
    }],
    steps: [{
      op: 'fill',
      target: locator('#search'),
      valueFrom: 'query',
      literalValue: null,
      key: null,
      expectNavigation: false,
      timeoutMs: 5000,
    }],
    output: {
      mode: 'page',
      collectionRoot: locator(),
      item: locator(),
      limit: 10,
      fields: [],
    },
    annotations: {
      readOnlyHint: true,
      untrustedContentHint: true,
    },
  },
  confidence: 0.9,
  evidence: 'Recorded once',
  issues: [],
});

test('validates a deterministic learned adapter', () => {
  assert.equal(validateManifest(adapter()).valid, true);
});

test('rejects an unknown schema version', () => {
  const candidate = adapter();
  candidate.schemaVersion = 'learned-adapter/99';
  const validation = validateManifest(candidate);
  assert.equal(validation.valid, false);
  assert.match(validation.errors.join(' '), /schemaVersion/);
});

test('builds WebMCP input schema and respects route scope', () => {
  const candidate = adapter();
  assert.deepEqual(buildInputSchema(candidate.tool), {
    type: 'object',
    properties: {
      query: { type: 'string', description: 'Search query' },
    },
    required: ['query'],
    additionalProperties: false,
  });
  assert.equal(manifestMatchesLocation(candidate, 'https://example.com/search?q=microphone'), true);
  assert.equal(manifestMatchesLocation(candidate, 'https://example.com/cart'), false);
});

File: backend/internal/manifest/manifest_test.go
package manifest

import "testing"

func pointer(value string) *string {
	return &value
}

func validAdapter() Adapter {
	return Adapter{
		SchemaVersion: SchemaVersion,
		Usable:        true,
		Site:          Site{Origin: "https://example.com", RoutePatterns: []string{"/search"}},
		Tool: Tool{
			Name:        "search_products",
			Description: "Search products",
			Safety:      "read",
			Parameters: []Parameter{{
				Name: "query", Description: "Search query", Type: "string", Required: true,
			}},
			Steps: []Step{{
				Operation: "fill", Target: Locator{CSS: pointer("#search")},
				ValueFrom: pointer("query"), TimeoutMS: 5000,
			}},
			Output: Output{Mode: "page", Limit: 10},
		},
		Confidence: 0.9,
	}
}

func TestValidateAcceptsDeterministicAdapter(t *testing.T) {
	if err := validAdapter().Validate(); err != nil {
		t.Fatalf("expected valid adapter, got %v", err)
	}
}

func TestValidateRejectsUnknownParameter(t *testing.T) {
	adapter := validAdapter()
	adapter.Tool.Steps[0].ValueFrom = pointer("missing")
	if err := adapter.Validate(); err == nil {
		t.Fatal("expected adapter validation to fail")
	}
}
